// @wallet-standard/ui-registry/lib/cjs/UiWalletHandleRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWalletForHandle_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = exports.getWalletAccountForUiWalletAccount_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = exports.registerWalletHandle_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = void 0;
exports.registerWalletHandle = registerWalletHandle;
exports.getWalletAccountForUiWalletAccount = getWalletAccountForUiWalletAccount;
exports.getWalletForHandle = getWalletForHandle;
const errors_1 = require("@wallet-standard/errors");
const uiWalletHandlesToWallets = new WeakMap();
/**
 * Associate a `UiWallet` or `UiWalletAccount` object with a Wallet Standard `Wallet` in the
 * central registry.
 *
 * This method is intended for Wallet Standard UI library authors, or app authors building directly
 * on the UI primitives without a framework-specific adapter.
 */
function registerWalletHandle(uiWalletHandle, wallet) {
    uiWalletHandlesToWallets.set(uiWalletHandle, wallet);
}
/** @deprecated Use {@link registerWalletHandle} instead. */
exports.registerWalletHandle_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = registerWalletHandle;
/**
 * Get the underlying `WalletAccount` object associated with a `UiWalletAccount` UI object.
 *
 * This method is intended for Wallet Standard UI library authors, or app authors building APIs that
 * need to materialize account-based features from UI primitives.
 */
function getWalletAccountForUiWalletAccount(uiWalletAccount) {
    const wallet = getWalletForHandle(uiWalletAccount);
    const account = wallet.accounts.find(({ address }) => address === uiWalletAccount.address);
    if (!account) {
        const err = new errors_1.WalletStandardError(errors_1.WALLET_STANDARD_ERROR__REGISTRY__WALLET_ACCOUNT_NOT_FOUND, {
            address: uiWalletAccount.address,
            walletName: wallet.name,
        });
        (0, errors_1.safeCaptureStackTrace)(err, getWalletAccountForUiWalletAccount);
        throw err;
    }
    return account;
}
/** @deprecated Use {@link getWalletAccountForUiWalletAccount} instead. */
exports.getWalletAccountForUiWalletAccount_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = getWalletAccountForUiWalletAccount;
/**
 * Get the underlying `Wallet` object associated with a `UiWallet` or `UiWalletAccount` UI object.
 *
 * This method is intended for Wallet Standard UI library authors, or app authors building APIs that
 * need to materialize wallet-based features from UI primitives.
 */
function getWalletForHandle(uiWalletHandle) {
    const wallet = uiWalletHandlesToWallets.get(uiWalletHandle);
    if (!wallet) {
        const err = new errors_1.WalletStandardError(errors_1.WALLET_STANDARD_ERROR__REGISTRY__WALLET_NOT_FOUND);
        (0, errors_1.safeCaptureStackTrace)(err, getWalletForHandle);
        throw err;
    }
    return wallet;
}
/** @deprecated Use {@link getWalletForHandle} instead. */
exports.getWalletForHandle_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = getWalletForHandle;
