// @mysten/sui/dist/zklogin/signer.mjs
import { extractClaimValue } from "./jwt-utils.mjs";
import { getZkLoginSignature } from "./signature.mjs";
import { toZkLoginPublicIdentifier } from "./publickey.mjs";
import { Signer } from "../cryptography/keypair.mjs";

//#region src/zklogin/signer.ts
/**
* A transport- and provider-agnostic zkLogin signer.
*
* It wraps any ephemeral {@link Signer} and, for every signing operation, transforms the ephemeral
* signature into a zkLogin signature using the supplied proof inputs and `maxEpoch`. Because every
* signing method on {@link Signer} funnels through `signWithIntent`, overriding that single method is
* enough to cover `signTransaction`, `signPersonalMessage`, and `signAndExecuteTransaction`.
*/
var ZkLoginSigner = class extends Signer {
	#ephemeralSigner;
	#maxEpoch;
	#inputs;
	#legacyAddress;
	#client;
	#publicKey;
	constructor(options) {
		super();
		this.#ephemeralSigner = options.ephemeralSigner;
		this.#maxEpoch = options.maxEpoch;
		this.#inputs = options.inputs;
		this.#legacyAddress = options.legacyAddress;
		this.#client = options.client;
		if (options.address !== void 0) {
			const derived = this.#derivePublicKey().toSuiAddress();
			if (derived !== options.address) throw new Error(`zkLogin proof does not match the provided address (derived ${derived}, expected ${options.address}) — check the \`legacyAddress\` flag`);
		}
	}
	#derivePublicKey() {
		return this.#publicKey ??= toZkLoginPublicIdentifier(BigInt(this.#inputs.addressSeed), extractClaimValue(this.#inputs.issBase64Details, "iss"), {
			legacyAddress: this.#legacyAddress,
			client: this.#client
		});
	}
	/**
	* The single extension point: the ephemeral signature is produced with the requested intent and
	* then wrapped in a zkLogin signature.
	*/
	async signWithIntent(bytes, intent) {
		const { bytes: signedBytes, signature: userSignature } = await this.#ephemeralSigner.signWithIntent(bytes, intent);
		return {
			bytes: signedBytes,
			signature: getZkLoginSignature({
				inputs: this.#inputs,
				maxEpoch: this.#maxEpoch,
				userSignature
			})
		};
	}
	sign(_data) {
		throw new Error("ZkLoginSigner does not support signing directly. Use signTransaction or signPersonalMessage instead");
	}
	getKeyScheme() {
		return "ZkLogin";
	}
	getPublicKey() {
		return this.#derivePublicKey();
	}
};

//#endregion
export { ZkLoginSigner };
