// @mysten/walrus/dist/flows/write-blob.mjs
import { WalrusClientError } from "../error.mjs";
import { CertificateBcs, parseCertificateFromBase64 } from "../utils/bcs.mjs";
import { fromBase64, toBase64 } from "@mysten/bcs";
import { Transaction } from "@mysten/sui/transactions";

//#region src/flows/write-blob.ts
function createWriteBlobFlow(client, ctx, options) {
	const { resume } = options;
	let encodeState;
	let registerState;
	let uploadState;
	let certifyState;
	const resumeBlobObjectId = resume && "blobObjectId" in resume ? resume.blobObjectId : void 0;
	const resumeTxDigest = resume && "txDigest" in resume ? resume.txDigest : void 0;
	const resumeNonce = resume && "nonce" in resume && resume.nonce ? fromBase64(resume.nonce) : void 0;
	const resumeBlobId = resume ? resume.blobId : void 0;
	const encode = async () => {
		if (!encodeState) {
			const { blob } = options;
			const hasUploadRelay = ctx.hasUploadRelay();
			const metadata$1 = hasUploadRelay ? await client.computeBlobMetadata({
				bytes: blob,
				nonce: resumeNonce
			}) : await client.encodeBlob(blob);
			if (resumeBlobId && metadata$1.blobId !== resumeBlobId) throw new WalrusClientError(`Resume blobId mismatch: expected ${resumeBlobId}, got ${metadata$1.blobId}. The blob content may have changed.`);
			encodeState = {
				metadata: metadata$1,
				size: blob.length,
				data: hasUploadRelay ? blob : void 0
			};
		}
		const { metadata, size } = encodeState;
		return {
			step: "encoded",
			blobId: metadata.blobId,
			rootHash: toBase64(metadata.rootHash),
			unencodedSize: size,
			..."nonce" in metadata ? { nonce: toBase64(metadata.nonce) } : {}
		};
	};
	const register = ({ epochs, deletable, owner, attributes }) => {
		if (!encodeState) throw new Error("encode must be executed before calling register");
		const { metadata, size } = encodeState;
		const transaction = new Transaction();
		transaction.setSenderIfNotSet(owner);
		if (ctx.hasUploadRelay()) {
			const meta = metadata;
			transaction.add(client.sendUploadRelayTip({
				size,
				blobDigest: meta.blobDigest,
				nonce: meta.nonce
			}));
		}
		transaction.transferObjects([client.registerBlob({
			size,
			epochs,
			blobId: metadata.blobId,
			rootHash: metadata.rootHash,
			deletable,
			attributes
		})], owner);
		registerState = {
			...encodeState,
			deletable
		};
		return transaction;
	};
	const upload = async (options$1) => {
		if (!encodeState) throw new Error("encode must be executed before calling upload");
		const { metadata, data } = encodeState;
		const deletable = options$1?.deletable ?? registerState?.deletable ?? false;
		let blobObject;
		if (resumeBlobObjectId) blobObject = await client.getBlobObject(resumeBlobObjectId);
		else if (options$1?.digest) blobObject = await ctx.getCreatedBlob(options$1.digest);
		else throw new Error("Either resume.blobObjectId or upload digest must be provided");
		let certificate;
		if (ctx.hasUploadRelay()) {
			const meta = metadata;
			certificate = (await client.writeBlobToUploadRelay({
				blobId: metadata.blobId,
				blob: data,
				nonce: meta.nonce,
				txDigest: options$1?.digest ?? resumeTxDigest,
				blobObjectId: blobObject.id,
				deletable,
				encodingType: meta.metadata.encodingType,
				signal: options$1?.signal
			})).certificate;
		} else {
			const meta = metadata;
			const confirmations = await client.writeEncodedBlobToNodes({
				blobId: metadata.blobId,
				objectId: blobObject.id,
				metadata: meta.metadata,
				sliversByNode: meta.sliversByNode,
				deletable,
				signal: options$1?.signal
			});
			certificate = await client.certificateFromConfirmations({
				confirmations,
				blobId: metadata.blobId,
				blobObjectId: blobObject.id,
				deletable
			});
		}
		uploadState = {
			blobObject,
			blobId: metadata.blobId,
			deletable,
			certificate
		};
		const txDigest = options$1?.digest ?? resumeTxDigest;
		return {
			step: "uploaded",
			blobId: metadata.blobId,
			blobObjectId: blobObject.id,
			...txDigest ? { txDigest } : {},
			certificate: CertificateBcs.serialize(certificate).toBase64()
		};
	};
	const certify = () => {
		if (!uploadState) throw new Error("upload must be executed before calling certify");
		const { blobObject, blobId, deletable, certificate } = uploadState;
		const transaction = client.certifyBlobTransaction({
			certificate,
			blobId,
			blobObjectId: blobObject.id,
			deletable
		});
		certifyState = {
			blobObject,
			blobId,
			transaction
		};
		return transaction;
	};
	const executeRegister = async ({ signer, ...options$1 }) => {
		const transaction = register(options$1);
		const { digest } = await ctx.executeTransaction(transaction, signer, "register blob");
		return {
			step: "registered",
			blobId: encodeState.metadata.blobId,
			blobObjectId: (await ctx.getCreatedBlob(digest)).id,
			txDigest: digest,
			..."nonce" in encodeState.metadata ? { nonce: toBase64(encodeState.metadata.nonce) } : {}
		};
	};
	const executeCertify = async ({ signer }) => {
		const transaction = certify();
		await ctx.executeTransaction(transaction, signer, "certify blob");
		const blobObject = await ctx.loadBlobObject(uploadState.blobObject.id);
		return {
			step: "certified",
			blobId: uploadState.blobId,
			blobObjectId: blobObject.id,
			blobObject
		};
	};
	const getBlob = async () => {
		if (!certifyState && !uploadState) throw new Error("upload or certify must be executed before calling getBlob");
		const state = certifyState ?? uploadState;
		const blobObject = await ctx.loadBlobObject(state.blobObject.id);
		return {
			step: "certified",
			blobId: state.blobId,
			blobObjectId: blobObject.id,
			blobObject
		};
	};
	const restoreUploadState = async (resumeData, deletable) => {
		uploadState = {
			blobObject: await client.getBlobObject(resumeData.blobObjectId),
			blobId: resumeData.blobId,
			deletable,
			certificate: parseCertificateFromBase64(resumeData.certificate)
		};
	};
	/** @yields {WriteBlobStep} */
	async function* run(options$1) {
		const resumeStep = resume?.step;
		const stepOrder = [
			"encoded",
			"registered",
			"uploaded",
			"certified"
		];
		const resumeIndex = resumeStep ? stepOrder.indexOf(resumeStep) : -1;
		if (resumeIndex >= stepOrder.indexOf("certified")) return;
		if (resumeIndex < stepOrder.indexOf("encoded")) yield await encode();
		else if (resumeIndex < stepOrder.indexOf("uploaded")) await encode();
		let registerDigest;
		if (!resumeBlobObjectId) {
			const regResult = await executeRegister({
				signer: options$1.signer,
				epochs: options$1.epochs,
				deletable: options$1.deletable,
				owner: options$1.owner ?? options$1.signer.toSuiAddress(),
				attributes: options$1.attributes
			});
			registerDigest = regResult.txDigest;
			yield regResult;
		} else registerState = {
			...encodeState,
			deletable: options$1.deletable
		};
		if (resumeIndex < stepOrder.indexOf("uploaded")) yield await upload({
			digest: registerDigest ?? resumeTxDigest,
			signal: options$1.signal
		});
		else await restoreUploadState(resume, options$1.deletable);
		yield await executeCertify({ signer: options$1.signer });
	}
	return {
		encode,
		register,
		upload,
		certify,
		getBlob,
		executeRegister,
		executeCertify,
		run
	};
}

//#endregion
export { createWriteBlobFlow };
