// @mysten/walrus/dist/flows/write-files.mjs
import { encodeQuiltPatchId } from "../utils/quilts.mjs";

//#region src/flows/write-files.ts
function createWriteFilesFlow(client, { files, resume }) {
	let quiltBytes;
	let quiltIndex;
	const blobFlow = client.writeBlobFlow({
		get blob() {
			if (!quiltBytes) throw new Error("encode must be executed before accessing blob");
			return quiltBytes;
		},
		resume
	});
	const encode = async () => {
		if (!quiltBytes) {
			const { quilt, index } = await client.encodeQuilt({ blobs: await Promise.all(files.map(async (file, i) => ({
				contents: await file.bytes(),
				identifier: await file.getIdentifier() ?? `file-${i}`,
				tags: await file.getTags() ?? {}
			}))) });
			quiltBytes = quilt;
			quiltIndex = index;
		}
		return blobFlow.encode();
	};
	const register = (options) => {
		if (!quiltBytes) throw new Error("encode must be executed before calling register");
		return blobFlow.register({
			...options,
			attributes: {
				_walrusBlobType: "quilt",
				...options.attributes
			}
		});
	};
	const listFiles = async () => {
		if (!quiltIndex) throw new Error("encode must be executed before calling listFiles");
		const certResult = await blobFlow.getBlob();
		return quiltIndex.patches.map((patch) => ({
			id: encodeQuiltPatchId({
				quiltId: certResult.blobId,
				patchId: {
					version: 1,
					startIndex: patch.startIndex,
					endIndex: patch.endIndex
				}
			}),
			blobId: certResult.blobId,
			blobObject: certResult.blobObject
		}));
	};
	/** @yields {WriteBlobStep} */
	async function* run(options) {
		const resumeStep = resume?.step;
		const stepOrder = [
			"encoded",
			"registered",
			"uploaded",
			"certified"
		];
		const resumeIndex = resumeStep ? stepOrder.indexOf(resumeStep) : -1;
		if (resumeIndex >= stepOrder.indexOf("certified")) return;
		if (resumeIndex < stepOrder.indexOf("encoded")) yield await encode();
		else if (resumeIndex < stepOrder.indexOf("uploaded")) await encode();
		const resumeBlobObjectId = resume && "blobObjectId" in resume ? resume.blobObjectId : void 0;
		let registerDigest;
		if (!resumeBlobObjectId) {
			const regResult = await blobFlow.executeRegister({
				signer: options.signer,
				epochs: options.epochs,
				deletable: options.deletable,
				owner: options.owner ?? options.signer.toSuiAddress(),
				attributes: {
					_walrusBlobType: "quilt",
					...options.attributes
				}
			});
			registerDigest = regResult.txDigest;
			yield regResult;
		}
		if (resumeIndex < stepOrder.indexOf("uploaded")) yield await blobFlow.upload({
			digest: registerDigest,
			deletable: options.deletable,
			signal: options.signal
		});
		if (resumeIndex < stepOrder.indexOf("certified")) yield await blobFlow.executeCertify({ signer: options.signer });
	}
	return {
		encode,
		register,
		upload: blobFlow.upload,
		certify: blobFlow.certify,
		listFiles,
		executeRegister: async (options) => {
			const { signer, ...rest } = options;
			return blobFlow.executeRegister({
				signer,
				...rest,
				attributes: {
					_walrusBlobType: "quilt",
					...rest.attributes
				}
			});
		},
		executeCertify: blobFlow.executeCertify,
		run
	};
}

//#endregion
export { createWriteFilesFlow };
